rmw_fastrtps_shared_cpp/src/rmw_get_topic_info.cpp

```diff
@@ -0,0 +1,265 @@
+#include <map>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "rmw/rmw.h"
+#include "rmw/types.h"
+#include "rmw/topic_info_array.h"
+#include "rmw/topic_info.h"
+
+#include "rmw_fastrtps_shared_cpp/custom_participant_info.hpp"
+#include "rmw_fastrtps_shared_cpp/rmw_common.hpp"
+#include "rmw_fastrtps_shared_cpp/namespace_prefix.hpp"
+
+namespace rmw_fastrtps_shared_cpp
+{
+
+rmw_ret_t
+_validate_params(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  rmw_topic_info_array_t * participants_info)
+{
+  if (!identifier) {
+    RMW_SET_ERROR_MSG("null implementation identifier provided");
+    return RMW_RET_ERROR;
+  }
+
+  if (!topic_name) {
+    RMW_SET_ERROR_MSG("null topic_name provided");
+    return RMW_RET_ERROR;
+  }
+
+  if (!allocator) {
+    RMW_SET_ERROR_MSG("null allocator provided");
+    return RMW_RET_ERROR;
+  }
+
+  if (!node) {
+    RMW_SET_ERROR_MSG("null node handle");
+    return RMW_RET_ERROR;
+  }
+
+  // Get participant pointer from node
+  if (node->implementation_identifier != identifier) {
+    RMW_SET_ERROR_MSG("node handle not from this implementation");
+    return RMW_RET_ERROR;
+  }
+
+  if (!participants_info) {
+    RMW_SET_ERROR_MSG("null participants_info provided");
+    return RMW_RET_ERROR;
+  }
+
+  return RMW_RET_OK;
+}
+
+std::vector<std::string>
+_get_topic_fqdns(const char * topic_name, bool no_mangle)
+{
+  std::vector<std::string> topic_fqdns;
+  topic_fqdns.push_back(topic_name);
+  if (!no_mangle) {
+    auto ros_prefixes = _get_all_ros_prefixes();
+    // Build the list of all possible topic FQDN
+    std::for_each(ros_prefixes.begin(), ros_prefixes.end(),
+      [&topic_fqdns, &topic_name](const std::string & prefix) {
+        topic_fqdns.push_back(prefix + topic_name);
+      });
+  }
+  return topic_fqdns;
+}
+
+rmw_ret_t
+_set_rmw_topic_info(
+  rmw_topic_info_t * topic_info,
+  const GUID_t & participant_guid,
+  const char * node_name,
+  const char * node_namespace,
+  TopicData topic_data,
+  ::ParticipantListener * slave_target,
+  rcutils_allocator_t * allocator)
+{
+  static_assert(
+    sizeof(eprosima::fastrtps::rtps::GUID_t) <= RMW_GID_STORAGE_SIZE,
+    "RMW_GID_STORAGE_SIZE insufficient to store the rmw_fastrtps_cpp GID implementation."
+  );
+  uint8_t rmw_gid[RMW_GID_STORAGE_SIZE];
+  memset(&rmw_gid, 0, RMW_GID_STORAGE_SIZE);
+  memcpy(&rmw_gid, &topic_data.guid, sizeof(eprosima::fastrtps::rtps::GUID_t));
+  rmw_ret_t ret = rmw_topic_info_set_gid(topic_info, rmw_gid,
+      sizeof(eprosima::fastrtps::rtps::GUID_t));
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  // set topic type
+  ret = rmw_topic_info_set_topic_type(topic_info, topic_data.topic_type.c_str(), allocator);
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  // set qos profile
+  ret = rmw_topic_info_set_qos_profile(topic_info, &topic_data.qos_profile);
+  if (ret != RMW_RET_OK) {
+    return ret;
```

Comment:
You have to cleanup previous allocations before returning an error.
Same applies for later return statements in this function.

Reply:
Using `_handle_topic_info_member_fini` functions to do it see line 148